p2p/pex/pex_reactor.go

```diff
@@ -453,47 +453,91 @@ func (r *Reactor) ensurePeers() {
 	// not perfect, but somewhate ensures that we prioritize connecting to more-vetted
 	// NOTE: range here is [10, 90]. Too high ?
 	newBias := cmtmath.MinInt(out, 8)*10 + 10
-
 	toDial := make(map[p2p.ID]*p2p.NetAddress)
-	// Try maxAttempts times to pick numToDial addresses to dial
-	maxAttempts := numToDial * 3
 
-	for i := 0; i < maxAttempts && len(toDial) < numToDial; i++ {
-		if !r.IsRunning() || !r.book.IsRunning() {
-			return
-		}
+	// Calculate reserve size as 50% of numToDial with a minimum of 10
+	reserveSize := cmtmath.MaxInt(numToDial/2, 10)
 
-		try := r.book.PickAddress(newBias)
-		if try == nil {
-			continue
-		}
-		if _, selected := toDial[try.ID]; selected {
+	// Try maxAttempts times to pick numToDial + reserveSize addresses to dial
+	maxToDialAttempts := numToDial * 3
+
+	for i := 0; i < maxToDialAttempts && len(toDial) < numToDial+reserveSize; i++ {
+		prospectivePeers := r.book.GetSelectionWithBias(newBias)
+		if prospectivePeers == nil {
 			continue
 		}
-		if r.Switch.IsDialingOrExistingAddress(try) {
-			continue
+		for _, prospectivePeer := range prospectivePeers {
+			if r.filterAddress(prospectivePeer, toDial) {
+				toDial[prospectivePeer.ID] = prospectivePeer
+				if len(toDial) >= numToDial+reserveSize {
+					break
+				}
+			}
 		}
-		// TODO: consider moving some checks from toDial into here
-		// so we don't even consider dialing peers that we want to wait
-		// before dialing again, or have dialed too many times already
-		toDial[try.ID] = try
 	}
 
-	// Dial picked addresses
-	for _, addr := range toDial {
-		go func(addr *p2p.NetAddress) {
-			err := r.dialPeer(addr)
-			if err != nil {
-				switch err.(type) {
-				case ErrMaxAttemptsToDial, ErrTooEarlyToDial:
-					r.Logger.Debug(err.Error(), "addr", addr)
-				default:
+	toDialCount := len(toDial)
+	successCount := 0
+	errorCount := 0
+
+	for successCount < numToDial && len(toDial) > 0 {
```

Comment:
This for is now blocking, right? We don't leave `ensurePeers` until the conditions are observed, namely, we dialed everyone we were supposed to.